pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+								},
+							},
+						},
+						{
+							Name: "tls",
+							VolumeSource: corev1.VolumeSource{
+								Secret: &corev1.SecretVolumeSource{
+									SecretName: DataPlaneName,
+								},
+							},
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name:  "dataplane",
+							Image: r.instance.Spec.ContainerRegistry + DataplaneImage + ":" + r.instance.Spec.ImageTag,
+							Args: []string{
+								"--log-level", r.instance.Spec.LogLevel,
+								"--controlplane-host", ControlPlaneName,
+							},
+							ImagePullPolicy: corev1.PullIfNotPresent,
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: dpapi.ListenPort,
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "ca",
+									MountPath: dpapp.CAFile,
+									SubPath:   "ca",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "tls",
+									MountPath: dpapp.CertificateFile,
+									SubPath:   "cert",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "tls",
+									MountPath: dpapp.KeyFile,
+									SubPath:   "key",
+									ReadOnly:  true,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return r.createOrUpdateResource(ctx, dpDeployment)
+}
+
+// setService sets up a k8s service.
+func (r *InstanceReconciler) setService(ctx context.Context, name string, port uint16) error {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: r.clNamespace},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Protocol: corev1.ProtocolTCP,
+					Port:     int32(port),
+				},
+			},
+			Type:     corev1.ServiceTypeClusterIP,
+			Selector: map[string]string{"app": name},
+		},
+	}
+
+	return r.createResource(ctx, service)
+}
+
+// setService sets up k8s a persistent volume claim for the.
+func (r *InstanceReconciler) setPVC(ctx context.Context, name string) error {
+	// Create the PVC for cl-controlplane
+	controlplanePVC := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			Resources: corev1.VolumeResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource.MustParse("100Mi"),
+				},
+			},
+		},
+	}
+
+	return r.createResource(ctx, controlplanePVC)
+}
+
+// setRules sets up k8s ClusterRule and ClusterRoleBinding for the controlplane.
+func (r *InstanceReconciler) setRules(ctx context.Context, name string) error {
+	// Create the ClusterRole for the controlplane.
+	clusterRole := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"services"},
+				Verbs:     []string{"create", "delete", "update"},
```

Comment:
watch? get?

Reply:
In the new controller we will need it In the last one it was reactive to the CLI. (I will update the in the issue https://github.com/clusterlink-net/clusterlink/issues/275)